### ISSUE TITLE
Failing to provide values for enum-typed properties can silently produce corrupt Avro data

### ISSUE THREAD

**Issue by woodlee** [ch-robinson/dotnet-avro#82]
In integrating Chr.Avro with the Confluent stack and producing from POCOs, we've run across a case where Avro-encoded messages are corrupt when enum-typed properties are null. Specifically, if the enum-typed property in the POCO is _not_ a `Nullable` but no value is passed, the serialized bytes may either default to the first value of the enum, or not write any bytes for that property at all. This simplified code sample which was run against v5.0.1 demonstrates:

```
using System;
using Chr.Avro.Representation;
using Chr.Avro.Serialization;

namespace IssueExample
{
    public enum MyExplicitEnum
    {
        MyEnumOne = 2,
        MyEnumTwo = 3
    }

    public enum MyImplicitEnum
    {
        MyEnumOne,
        MyEnumTwo
    }

    public class TestRecord
    {
        public bool MyBool { get; set; }
        public MyExplicitEnum MyExplicitEnum { get; set; }
        public string MyString { get; set; }
        public MyImplicitEnum MyImplicitEnum { get; set; }
    }

    internal static class Program
    {
        private const string SchemaJson = @"
        {
            ""type"": ""record"",
            ""name"": ""TestRecord"",
            ""fields"": [
                {
                    ""name"": ""MyBool"",
                    ""type"":  ""boolean""
                },
                {
                    ""name"": ""MyExplicitEnum"",
                    ""type"": {
                        ""type"": ""enum"",
                        ""name"": ""MyExplicitEnum"",
                        ""symbols"": [
                            ""MyEnumOne"",
                            ""MyEnumTwo""
                        ]
                    }
                },
                {
                    ""name"": ""MyString"",
                    ""type"": ""string""
                },
                {
                    ""name"": ""MyImplicitEnum"",
                    ""type"": {
                        ""type"": ""enum"",
                        ""name"": ""MyImplicitEnum"",
                        ""symbols"": [
                            ""MyEnumOne"",
                            ""MyEnumTwo""
                        ]
                    }
                }
            ]
        }
        ";

        private static void Main()
        {
            var schema = new JsonSchemaReader().Read(SchemaJson);
            var serializer = new BinarySerializerBuilder().BuildSerializer<TestRecord>(schema);

            var enumsSpecified = new TestRecord
            {
                MyBool = true,
                MyExplicitEnum = MyExplicitEnum.MyEnumTwo,
                MyString = "abcd",
                MyImplicitEnum = MyImplicitEnum.MyEnumTwo
            };

            // This prints 01-02-08-61-62-63-64-02... all OK:
            Console.WriteLine(BitConverter.ToString(serializer.Serialize(enumsSpecified)));

            var enumsOmitted = new TestRecord
            {
                MyBool = true,
                MyString = "abcd"
            };

            // But this prints 01-08-61-62-63-64-00... MyImplicitEnum has been "defaulted" to 00 (i.e. "MyEnumOne") and
            // MyExplicitEnum is not encoded at all, leading to corrupt data:
            Console.WriteLine(BitConverter.ToString(serializer.Serialize(enumsOmitted)));

            // If the JSON Avro schema is changed to make the two enumeration fields union types of [null, <the enum type>],
            // and we run again we get similar troubles:
            //   enumsSpecified: 01-02-02-08-61-62-63-64-02-02
            //   enumsOmitted:   01-02-08-61-62-63-64-02-00

            // Only if the types of the enums in the `TestRecord` POCO are made nullable does it behave as expected:
            //   Nullable in POCO, NOT nullable in Avro schema:
            //     Throws "System.InvalidOperationException: The binary operator Equal is not defined for the types
            //     'System.Nullable`1[IssueExample.MyExplicitEnum] and 'IssueExample.MyExplicitEnum'."
            //   Nullable in POCO, nullable in Avro schema:
            //     enumsSpecified: 01-02-02-08-61-62-63-64-02-02
            //     enumsOmitted:   01-00-08-61-62-63-64-00
        }
    }
}
```

When a non-`Nullable` property is left null on a POCO passed to Chr.Avro, I think I would expect to see a null reference exception instead. Does that seem reasonable? In the meantime we're just going to make the problem fields nullable in the POCOs we are passing to Chr.Avro.

**Comment by dstelljes** [comment]
Thanks for the report (and all the details)!

One note about your proposed solution: Enums are integers under the hood, so they default to 0 and cannot be null. In the `MyImplicitEnum` case, there's no way for the serializer to tell whether the record property was set to `MyEnumOne` or never touched.

The root cause here is the lack of a default case in the generated switch expression. Fortunately, the bug doesn't cause problems with legit enum values that don't have a match; that's checked when the serializer is built. I think the most correct solution would be to add a default case throwing `ArgumentOutOfRangeException` for any unnamed enum value.

**Comment by dstelljes** [comment]
Fix released with 5.0.2; also cherry picked into 2.1.1, 3.3.1, and 4.1.1 since it's a data corruption issue.

**Comment by woodlee** [comment]
Thank you so much, yet again--you're definitely making our lives easier! 😄
